client/src/controllers/ZoneController.ts

```diff
@@ -0,0 +1,44 @@
+import { Controller } from '@hotwired/stimulus';
+
+/**
+ * Enables the controlled element to respond to specific user interactions
+ * by adding or removing CSS classes dynamically.
+ *
+ * @example
+ * ```html
+ * <div
+ *   data-controller="w-zone"
+ *   data-w-zone-active-class="hovered active"
+ *   data-action="dragover->w-zone#activate dragleave->w-zone#deactivate"
+ * >
+ *   Drag files here
+ * </div>
+ * ```
+ * In above example:
+ * - The `hovered` and `active` classes are applied on `dragover`.
+ * - The classes are removed on `dragleave`.
+ */
+export class ZoneController extends Controller<HTMLElement> {
+  static classes = ['active'];
+
+  /**
+   * `activeClasses` is an array of CSS class names to be:
+   * - Applied when `event->w-zone#activate` is triggered.
+   * - Removed when `event->w-zone#deactivate` is triggered.
+   * These class names are defined in the `data-w-zone-active-class` attribute
+   * and should be space-separated.
+   */
+  declare activeClasses: string[];
+
+  connect() {
+    this.dispatch('ready', { cancelable: false });
+  }
+
```

Comment:
i agree on as the previous approach was straight forward and less leveraged, this is the most leveraged thing we can implement, i can not find any other better approach for this in the documentation, but one thing we can do here that what if we use active instead mod and it will be type bool because as far as i know we are not going to change this directly in DOM.

```
static values = {
    active: { type: Boolean, default: false },
  };

modeValueChanged(isActive) {
    if (isActive) {
      this.element.classList.add(...this.activeClasses);
    } else {
      this.element.classList.remove(...this.activeClasses);
    }
  }

  activate() {
    this.activeValue = true;
  }

  deactivate() {
    this.activeValue = false;
  }
```

'active' value could be conflict/confusion so we can use mod/effect name instead but 'active' works fine too i verified myself.

Reply:
@lb-  i found really unexpected behavior when testing cases and is failing with this implementation, 

test case for add classes is failing:

### ZoneController:

```
import { Controller } from '@hotwired/stimulus';

export class ZoneController extends Controller<HTMLElement> {
  static classes = ['active'];

  static values = {
    mode: { type: String, default: '' },
  };

  declare modeValue: string;

  declare activeClasses: string[];

  modeValueChanged(current, previous) {
    console.log(`-current: ${current}-previous: ${previous}-`)
    if (current == 'active') {
      console.log("classes ::: ",...this.activeClasses)
      this.element.classList.add(...this.activeClasses);
      console.log("Element classes ::: ",this.element.classList)
    } else {
      this.element.classList.remove(...this.activeClasses);
    }
  }
  
  activate() {
    this.modeValue = 'active';
  }

  deactivate() {
    this.modeValue = '';
  }

  noop() {}
}
```
(removed debounce for here)
</br>

### test case:

```
import { Application } from '@hotwired/stimulus';
import { ZoneController } from './ZoneController';

describe('ZoneController', () => {
  let app;

  const setup = async (html) => {
    document.body.innerHTML = `<main>${html}</main>`;

    app = Application.start();
    app.register('w-zone', ZoneController);

    await Promise.resolve();
  };

  afterEach(() => {
    app?.stop();
    jest.clearAllMocks();
  });

  describe('activate method', () => {
    it('should add active class to the element', async () => {
      await setup(`
        <div
          class="drop-zone"
          data-controller="w-zone"
          data-w-zone-active-class="hovered"
          data-action="dragover->w-zone#activate"
        ></div>
      `);

      const element = document.querySelector('.drop-zone');
      element.dispatchEvent(new Event('dragover'));

      expect(element.classList.contains('hovered')).toBe(true);
    });
  });
});
```
(this is not exact test cases, just tweak for where (when adding class) the failing is happening )

test case output: https://pastebin.com/pB1dZXag

**Note: I also tried previous implementation where test case passed smoothly**

one more, this thing is not happening in browser:

![image](https://github.com/user-attachments/assets/21ba47c9-711c-471f-aac4-c10048bebebd)

classes is being appended in browser so this, when happening in jest environment that state change and it does trigger callback but not adding classes (can be seen in [test case output](https://pastebin.com/pB1dZXag)) 

![image](https://github.com/user-attachments/assets/48ce1fe0-3d2c-4a45-944f-788756fbd8e5)

it already too late here i should go bed, it took long for me 🙂

Reply:
I debugged and researched for while long and then thought that assertion is happening before DOM changes as we started using DOM manipulation that will happened on state change so i added delay before assertion `await new Promise((resolve)=>setTimeout(resolve,0))` and yes tests was started pass 

And then i move forward for changes for debounce realized `await jest.runAllTimersAsync();` this will also be work as delay and yes, this made test pass you already mentioned this as well: 
https://github.com/wagtail/wagtail/pull/12569/files#r1842181556
but i thought it was just for debounce purpose only, i only wrote testing cases once before this.🥲

the biggest takeaway from this experience is to never rely too heavily on console.log for debugging.